### ISSUE TITLE
Potentially conflicting matrix option identifier strings

### ISSUE THREAD

**Issue by briansandall** [cubecart/v6#918]
This can occur if one has multiple products that use the same options, for example several different types of clothing that all come in the same sizes and colors.

Usually the product id is used in conjunction with the options_identifier string when querying the database, but if for whatever reason it is unavailable, the result would potentially be erroneous.

Pre- or post-fixing the string with the product_id should result in every string being unique, so long as it is separated by a period or some other separator to prevent possible matches between a multi-digit product_id and any option:value pair (e.g. product_id=13 could be confused as option_id=1 + value_id=3).

**Comment by bhsmither** [comment]
A product without a product_id won't happen.

A query where the needed product_id is missing is a mistake in the programming and **that** needs to be fixed.

Is there a case that could be made where the product_id is intentionally not included? Or have you found a section of code where the product_id is not checked for its presence prior to making a query?

**Comment by bhsmither** [comment]
Many moons ago, I've identified a problem with the options_identifier, and have brought that to the attention of the programmers. But it's been dismissed.

**Comment by briansandall** [comment]
I only noticed it when I was hacking around with a script to auto-generate and update matrix product codes based on the product options, and it was because I hadn't included the product_id in the query.

Whether there are any areas within CubeCart itself where this is truly an issue I don't know - haven't found any yet - but it could be really nice in some cases to be able to find the product_id if all you happen to have is the options_identifier.

Basically, I don't see any downside to making them truly unique, other than perhaps backwards compatibility (which is not a small downside if it becomes an issue).

What was the issue you raised before?

**Comment by bhsmither** [comment]
It's been said that working backwards from the options_identifier string is impossible. Personally, I would say things are not that bad -- maybe just *super incredibly difficult*.

I had explained to the programmers that string collision is just waiting to happen.

The string is composed of an undelimited concatenated series of **option_id** and **value_id** (if I recall). With an option_id of 1 and another option_id of 11, then a value_id of 11 and another value_id of 1, there ends up the sequence of 111 for both instances.

**Comment by briansandall** [comment]
Yep, that's how it works. The chances of collisions happening may be miniscule, but with enough products with enough options over enough stores, it will happen.

Collisions are the real concern; working backwards from a set of options is just what I was doing at the time I noticed it.

**Comment by briansandall** [comment]
Also, in my own store I already have collisions because I have multiple products that use the exact same set of options, so there is e.g. '1122' for both product ID 1 and 2.

I suppose that a collision would only be a concern if it could happen within a single product, since the product_id will (should) always be a condition in the query's `WHERE` clause.

**Comment by abrookbanks** [comment]
This is easy to resolve by adding the product id but I don't know how to handle existing data.

**Comment by abrookbanks** [comment]
> I already have collisions because I have multiple products that use the exact same set of options

I search the code for allCubeCart_option_matrix SQL queries. It doesn't matter if there are collisions on duplicate hash values as all queries contain a where clause on the CubeCart_option_matrix.product_id column. 

In conclusion nothing that I can see of needs to be changed.
